### PR TITLE
Fix index out of range error when data not found

### LINKED ISSUE #437: Improper IndexError handling
`File "/home/..../lib/python3.11/site-packages/pypistats/__init__.py", line 261, in _date_range
    first = data[0]["date"]
            ~~~~^^^
IndexError: list index out of range`

The exception does not handle the data list is empty.
The error will cause the program to stop without proper handling.

### PR DESCRIPTION
**Pull request by Yao-Wen-Chang**
close #437

### CODE REVIEW

**Comment by Yao-Wen-Chang** [comment]
@hugovk What are those failures in the CI process?
How to fix those issues?

**Comment by hugovk** [comment]
Running against a project which [doesn't exist on PyPI](https://pypi.org/project/testbrojct/) and getting results in JSON format: 

```console
❯ pypistats python_major testbrojct -f json
{"data": [], "package": "testbrojct", "type": "python_major_downloads"}
```

However, we still get the error when asking for results in other formats:

```pytb
❯ pypistats python_major testbrojct
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.12/bin/pypistats", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/hugo/github/pypistats/src/pypistats/cli.py", line 372, in main
    args.func(args)
  File "/Users/hugo/github/pypistats/src/pypistats/cli.py", line 258, in python_major
    pypistats.python_major(
  File "/Users/hugo/github/pypistats/src/pypistats/__init__.py", line 497, in python_major
    return pypi_stats_api(endpoint, params, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hugo/github/pypistats/src/pypistats/__init__.py", line 169, in pypi_stats_api
    data = _percent(data)
           ^^^^^^^^^^^^^^
  File "/Users/hugo/github/pypistats/src/pypistats/__init__.py", line 322, in _percent
    grand_total = _grand_total_value(data)
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hugo/github/pypistats/src/pypistats/__init__.py", line 282, in _grand_total_value
    if data[0]["category"] in ["with_mirrors", "without_mirrors"]:
       ~~~~^^^
IndexError: list index out of range
```
```pytb
❯ pypistats python_major testbrojct -f html
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.12/bin/pypistats", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/hugo/github/pypistats/src/pypistats/cli.py", line 372, in main
    args.func(args)
  File "/Users/hugo/github/pypistats/src/pypistats/cli.py", line 258, in python_major
    pypistats.python_major(
  File "/Users/hugo/github/pypistats/src/pypistats/__init__.py", line 497, in python_major
    return pypi_stats_api(endpoint, params, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hugo/github/pypistats/src/pypistats/__init__.py", line 169, in pypi_stats_api
    data = _percent(data)
           ^^^^^^^^^^^^^^
  File "/Users/hugo/github/pypistats/src/pypistats/__init__.py", line 322, in _percent
    grand_total = _grand_total_value(data)
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hugo/github/pypistats/src/pypistats/__init__.py", line 282, in _grand_total_value
    if data[0]["category"] in ["with_mirrors", "without_mirrors"]:
       ~~~~^^^
IndexError: list index out of range
```

Let's also handle these too. And please can you add some tests?

For example, see https://github.com/hugovk/pypistats/pull/73/commits/a43357deaf65f9d2a2e7698727aeada180be247f that added the `KeyError` handling with tests.

**Comment by Yao-Wen-Chang** [comment]
I will fix those bugs and add the test.

File: src/pypistats/__init__.py [hunk 1]
```diff
@@ -261,7 +264,7 @@ def _date_range(data: dict | list) -> tuple[str | None, str | None]:
     try:
         first = data[0]["date"]
         last = data[0]["date"]
-    except KeyError:
+    except (KeyError, IndexError):
```
Comment (hugovk) [inline src/pypistats/__init__.py:267]: Is this no longer needed?
```suggestion
    except KeyError:
```

File: tests/test_pypistats.py [hunk 2]
```diff
@@ -747,3 +747,10 @@ def test_format_pandas(self) -> None:
         # Assert
         assert isinstance(output, pandas.DataFrame)
         assert str(output).strip() == expected_output.strip()
+
+    def test__package_not_exist(self) -> None:
```
Comment (hugovk) [inline tests/test_pypistats.py:751]:
```suggestion
    def test_package_not_exist(self) -> None:
```

File: tests/test_pypistats.py [hunk 3]
```diff
@@ -747,3 +747,10 @@ def test_format_pandas(self) -> None:
...
+        package = "a" * 100
+        result = pypistats.python_major(package)
```
Comment (hugovk) [inline tests/test_pypistats.py:753]: Rather than making network calls and hitting the real API, please can you mock this, like the other tests?

**Review by hugovk — CHANGES_REQUESTED** [review]

File: tests/test_pypistats.py (hunk 3 above)
Comment (Yao-Wen-Chang) [reply tests/test_pypistats.py:753]: Sure!

File: src/pypistats/__init__.py (hunk 1 above)
Comment (Yao-Wen-Chang) [reply src/pypistats/__init__.py:267]: Yep, I'll remove it.

**Comment by Yao-Wen-Chang** [comment]
Do you know why using the 
`Command: pypistats python_major -j aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa`
it returns 200, but
`Command: pypistats overall -j aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa`
it returns 404 from the API call, then raises the exception.

Reasonably, both should return 404, because the package does not exist.

File: tests/test_pypistats.py (hunk 2 above)
Comment (Yao-Wen-Chang) [reply tests/test_pypistats.py]: Done

File: src/pypistats/__init__.py (hunk 1 above)
Comment (Yao-Wen-Chang) [reply src/pypistats/__init__.py]: Done

**Comment by hugovk** [comment]
I don't know.

If we run with `--verbose` it shows the API endpoints:

* https://pypistats.org/api/packages/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/python_major

* https://pypistats.org/api/packages/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/overall

As you say, the first returns 200 with empty `data`, the second is 404.

Turns out I already reported this upstream 3 years ago, but there's been no response so we'll have to live with it:

* https://github.com/crflynn/pypistats.org/issues/43

File: src/pypistats/__init__.py [hunk 4]
```diff
@@ -124,6 +124,9 @@ def pypi_stats_api(
 
         _save_cache(cache_file, res)
 
+    if not res.get("data", []):
+        return f"The package '{res.get('package', '')}' does not exist"
```
Comment (hugovk) [inline src/pypistats/__init__.py:128]: This also happens when looking up a new package that hasn't had any downloads yet. I think we won't see any numbers until the next day. 

For example, looking at the new packages feed: https://pypi.org/rss/packages.xml

https://pypi.org/project/hnt-sap-service-library/ is a new package with no downloads registered yet and also triggers this. 

Shall we update the message?

```suggestion
        return f"No data found for '{res.get('package', '')}'"
```

Or maybe link to it:

```suggestion
        return f"No data found for https://pypi.org/project/{res.get('package', '')}/"
```

File: tests/test_pypistats.py [hunk 5]
```diff
@@ -747,3 +747,22 @@ def test_format_pandas(self) -> None:
...
+    def test_package_not_exist(self) -> None:
+        # Arrange
+        package = "a" * 100
+        mocked_response = f"""{{
+            "data":[],
+            "package":"{package}",
+            "type":"python_major_downloads"
+        }}"""
+        mocked_url = f"https://pypistats.org/api/packages/{package}/python_major"
+        expected_output = f"The package '{package}' does not exist"
```
Comment (hugovk) [inline tests/test_pypistats.py:761]:
```suggestion
        expected_output = f"No data for {package}'"
```
or
```suggestion
        expected_output = f"No data for https://pypi.org/project/{package}/"
```

File: src/pypistats/__init__.py (hunk 4 above)
Comment (Yao-Wen-Chang) [reply src/pypistats/__init__.py:128]: You are right. I noticed this issue. 
I think there should be some downloads on the first day from mirrors. 
I reckon this tool is still very useful. 
I prefer the second one. Great suggestion!

**Comment by Yao-Wen-Chang** [comment]
For this line in the code: https://github.com/hugovk/pypistats/blob/bfda2c984f05adfc794a11744db621476a71a836/src/pypistats/__init__.py#L121
Do you think it would be a good idea to delete it, considering it raises an error message?

**Review by hugovk — APPROVED** [review]

**Comment by hugovk** [comment]
Thank you!

**Pull request merged**